astropy/io/fits/column.py

```diff
@@ -1521,6 +1523,11 @@ def _update_column_attribute_changed(self, column, attr, old_value,
             if col is column:
                 break
 
+        if attr == 'name':
+            del self.names
+        elif attr == 'format':
+            del self.formats
```

Comment:
Doesn't quite work--these also need to be invalidated when columns are added/removed.  Otherwise I'm fine with the general concept.  My preference would still be to not make any changes in the `ColDefs` class and instead update the `FITS_rec` class to be less reliant on the `ColDefs` object in general (an object I would like to get rid of, ultimately).  

That said, in the interest of not wasting any more of your time on this and getting it finished, if you can just fix the column add/remove bug and ensure all the tests are passing this is definitely good enough for now! :+1: 

I also noticed you cleaned up a little bit of additional code rot--thanks for that.